File: app/src/main/java/com/ytempest/wanandroid/activity/architecture/content/ArchArticlePresenter.java
package com.ytempest.wanandroid.activity.architecture.content;

import androidx.annotation.NonNull;

import com.ytempest.wanandroid.base.presenter.BasePresenter;
import com.ytempest.wanandroid.http.bean.ArchitectureContentBean;
import com.ytempest.wanandroid.http.bean.ArticleCollectBean;
import com.ytempest.wanandroid.http.bean.BaseResp;
import com.ytempest.wanandroid.http.observer.HandlerObserver;
import com.ytempest.wanandroid.interactor.impl.BaseInteractor;
import com.ytempest.wanandroid.utils.PageCtrl;
import com.ytempest.wanandroid.utils.RxUtils;

import javax.inject.Inject;

import io.reactivex.Observable;

/**
 * @author heqidu
 * @since 2021/1/6
 */
public class ArchArticlePresenter extends BasePresenter<IArchArticleContract.View> implements IArchArticleContract.Presenter {

    private static final String TAG = ArchArticlePresenter.class.getSimpleName();

    private final PageCtrl mPageCtrl = new PageCtrl();

    @Inject
    public ArchArticlePresenter(BaseInteractor interactor) {
        super(interactor);
    }

    @Override
    public boolean isUserLogin() {
        return mInteractor.getConfigs().getUser().isUserLogin();
    }

    public void refreshArchitectureContent(int id) {
        if (mPageCtrl.isRequesting()) return;
        mPageCtrl.moveTo(PageCtrl.State.REFRESH);
        mInteractor.getHttpHelper().getArchitectureContent(id, mPageCtrl.getNextPage())
                .compose(RxUtils.switchScheduler())
                .subscribe(new HandlerObserver<ArchitectureContentBean>(mView) {
                    @Override
                    protected void onSuccess(@NonNull ArchitectureContentBean architectureContentBean) {
                        mPageCtrl.moveTo(PageCtrl.State.SUCCESS);
                        mView.onLoadArchitectureContent(architectureContentBean, true);
                    }

                    @Override
                    protected void onFail(int code, Throwable e) {
                        super.onFail(code, e);
                        mPageCtrl.moveTo(PageCtrl.State.FAIL);
                        mView.onRefreshArchitectureFail(code);
                    }
                });
    }

    @Override
    public void loadMoreArchitectureContent(int id) {
        if (mPageCtrl.isRequesting()) return;
        mPageCtrl.moveTo(PageCtrl.State.LOAD_MORE);
        mInteractor.getHttpHelper().getArchitectureContent(id, mPageCtrl.getNextPage())
                .compose(RxUtils.switchScheduler())
                .subscribe(new HandlerObserver<ArchitectureContentBean>(mView) {
                    @Override
                    protected void onSuccess(@NonNull ArchitectureContentBean architectureContentBean) {
                        mPageCtrl.moveTo(PageCtrl.State.SUCCESS);
                        mView.onLoadArchitectureContent(architectureContentBean, false);
                    }

                    @Override
                    protected void onFail(int code, Throwable e) {
                        super.onFail(code, e);
                        mPageCtrl.moveTo(PageCtrl.State.FAIL);
                    }
                });
    }

    @Override
    public void updateArticleCollectStatus(boolean isCollect, ArchitectureContentBean.DatasBean article) {
        Observable<BaseResp<ArticleCollectBean>> collectObservable;
        if (isCollect) { // 收藏文章
            collectObservable = mInteractor.getHttpHelper().addCollectArticle(article.getId());
        } else { // 取消收藏
            collectObservable = mInteractor.getHttpHelper().cancelCollectArticle(article.getId());
        }

        collectObservable.compose(RxUtils.switchScheduler())
                .map(RxUtils.checkArticleCollectData())
                .subscribe(new HandlerObserver<ArticleCollectBean>(mView, HandlerObserver.SHOW_LOADING) {
                    @Override
                    protected void onSuccess(@NonNull ArticleCollectBean bean) {
                        mView.onArchArticleCollectSuccess(isCollect, article);
                    }

                    @Override
                    protected void onFail(int code, Throwable e) {
                        super.onFail(code, e);
                        mView.onArchArticleCollectFail(isCollect, article, code);
                    }
                });
    }
}

File: app/src/main/java/com/ytempest/wanandroid/activity/architecture/content/ArchArticlePresenter.kt
package com.ytempest.wanandroid.activity.architecture.content

import com.ytempest.wanandroid.base.presenter.BasePresenter
import com.ytempest.wanandroid.http.observer.HandlerObserver
import com.ytempest.wanandroid.interactor.impl.BaseInteractor
import com.ytempest.wanandroid.utils.RxUtils
import com.ytempest.wanandroid.http.bean.ArchitectureContentBean
import com.ytempest.wanandroid.http.bean.ArticleCollectBean
import com.ytempest.wanandroid.utils.PageCtrl
import javax.inject.Inject

/**
 * @author heqidu
 * @since 21-2-22
 */
class ArchArticlePresenter @Inject constructor(
        interactor: BaseInteractor
) : BasePresenter<IArchArticleView>(interactor), IArchArticlePresenter {

    private val mPageCtrl = PageCtrl()

    override fun isUserLogin(): Boolean {
        return mInteractor.getConfigs().getUser().isUserLogin()
    }

    override fun refreshArchitectureContent(id: Int) {
        if (mPageCtrl.isRequesting) return
        mPageCtrl.moveTo(PageCtrl.State.REFRESH)
        mInteractor.getHttpHelper().getArchitectureContent(id, mPageCtrl.nextPage)
                .compose(RxUtils.switchScheduler())
                .subscribe(object : HandlerObserver<ArchitectureContentBean>(mView) {
                    override fun onSuccess(bean: ArchitectureContentBean) {
                        mPageCtrl.moveTo(PageCtrl.State.SUCCESS)
                        mView.onLoadArchitectureContent(bean, true)
                    }

                    override fun onFail(code: Int, e: Throwable) {
                        super.onFail(code, e)
                        mPageCtrl.moveTo(PageCtrl.State.FAIL)
                        mView.onRefreshArchitectureFail(code)
                    }
                })
    }

    override fun loadMoreArchitectureContent(id: Int) {
        if (mPageCtrl.isRequesting) return
        mPageCtrl.moveTo(PageCtrl.State.LOAD_MORE)
        mInteractor.getHttpHelper().getArchitectureContent(id, mPageCtrl.nextPage)
                .compose(RxUtils.switchScheduler())
                .subscribe(object : HandlerObserver<ArchitectureContentBean>(mView) {
                    override fun onSuccess(architectureContentBean: ArchitectureContentBean) {
                        mPageCtrl.moveTo(PageCtrl.State.SUCCESS)
                        mView.onLoadArchitectureContent(architectureContentBean, false)
                    }

                    override fun onFail(code: Int, e: Throwable) {
                        super.onFail(code, e)
                        mPageCtrl.moveTo(PageCtrl.State.FAIL)
                    }
                })
    }

    override fun updateArticleCollectStatus(isCollect: Boolean, article: ArchitectureContentBean.Data) {
        val collectObservable = if (isCollect) mInteractor.getHttpHelper().addCollectArticle(article.id.toLong()) // 收藏文章
        else mInteractor.getHttpHelper().cancelCollectArticle(article.id.toLong()) // 取消收藏

        collectObservable.compose(RxUtils.switchScheduler())
                .map(RxUtils.checkArticleCollectData())
                .subscribe(object : HandlerObserver<ArticleCollectBean>(mView, Flags.SHOW_LOADING) {
                    override fun onSuccess(bean: ArticleCollectBean) {
                        mView.onArchArticleCollectSuccess(isCollect, article)
                    }

                    override fun onFail(code: Int, e: Throwable) {
                        super.onFail(code, e)
                        mView.onArchArticleCollectFail(isCollect, article, code)
                    }
                })
    }
}